Integrate OpenAI Function Calling API

MorphlingEd opened the pull request:
## Description

- Introduced a new `FuncConfig` in `config.py`, including an extra field to add function information objects (`Dict` objects).
- Implemented a basic docstring parser in `util/functions`, which is used to parse docstrings of functions into dictionaries to be passed to LMs.
- Added new data classes for packing functions (currently only basic integer algebra and Wiki-based search)
- Added new parameters in `ChatAgent` to control whether function calling is enabled. When function calling is enabled, information about any called functions will be recorded into `FuncRecord` objects (output in `info`) and added to messages (in `FuncMessage`)

`examples/function_call/role_playing_with_function.py` has a basic demo, which can print out the information of the called functions in each step. 
https://drive.google.com/file/d/1U8CoqPBYWrXNwvRT-8QvK8vXtjP5_rnf/view?usp=sharing

Introduced quite many new classes and hope you could provide suggestions for any changes.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
close #187
- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues/186) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

Obs01ete commented on camel/configs.py:
```diff
@@ -74,3 +74,71 @@ class ChatGPTConfig:
+
+
+@dataclass
+class FuncConfig:
```
This is an extraordinary code duplication. Not good to copy-paste a huge chunk of code and change one line. Please rework with either composition or inheritance.

Obs01ete commented on camel/societies/role_playing.py:
```diff
@@ -400,6 +425,14 @@ def step(
             whether the user agent terminated the conversation, and any
             additional user information.
         """
+        # print("Assistant:")
```
Remove all commented out code.

Obs01ete replied:
Still not resolved

Obs01ete commented on examples/function_call/role_playing_with_function.py:
```diff
@@ -0,0 +1,87 @@
+
+
+def main(model_type=ModelType.GPT_4_FUNC) -> None:
+    task_prompt = ("Estimate KAUST's current age by searching the year when"
```
Nice testing

Obs01ete commented on examples/function_call/role_playing_with_function.py:
```diff
@@ -0,0 +1,87 @@
+from camel.utils import print_text_animated
+
+
+def main(model_type=ModelType.GPT_4_FUNC) -> None:
```
Add a test for the example code

MorphlingEd replied:
sorry about this and I will do this :)

Obs01ete commented on camel/messages/func_message.py:
```diff
@@ -0,0 +1,97 @@
+        msg_dict = {"role": "assistant"}
+        msg_dict["content"] = self.content
+
+        msg_dict["function_call"] = {}
```
Same here. Pre-construct a function_call dict and feed it into the constructor of msg_dict.

MorphlingEd replied:
That's a truly good point. Will modify this later.

Obs01ete reviewed (changes requested):
It is cool functionality overall. However the code must be cleaned up.

dandansamax commented on camel/utils/functions.py:
```diff
@@ -239,3 +240,137 @@ def download_tasks(task: TaskType, folder_path: str) -> None:
+#     return func_dict
+
+
+def parse_doc(func: Callable) -> Dict[str, Any]:
```
Cool feature.

dandansamax commented on camel/functions/base.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@dataclass
+class BaseFuncs:
```
Moreover, the current `doc_parser` can only solve one layer parameters. We should have a method to contruct more complicated functions with python objects in the parameters. It is hard to generate by a doc but we can contruct it manually.

dandansamax replied:
I will write a example tonight to show my thoughts. Please checkout other comments firstly.

dandansamax reviewed (commented):
Overall, this is a very cool implementation! Just request a few more features.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -92,6 +119,10 @@ class ChatAgent(BaseAgent):
         agent. (default: :obj:`None`)
+        func_enable (bool): Whether to enable function calling
+            (default: :obj:`False`)
+        func_collects (Optional[List[BaseFuncs]]): List of function
```
Actually this is a list of function collection objects, where `BaseFuncs` is the base class of dataclasses containing functions. So I just assumed each object of `BaseFuncs` as one collection of functions and this argument should be a "collections"

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -112,17 +145,65 @@ def __init__(
                                  ModelType.GPT_3_5_TURBO)
-        self.model_config: ChatGPTConfig = model_config or ChatGPTConfig()
         self.message_window_size: Optional[int] = message_window_size
+        self.model_config_dict: dict
```
This was because when function calling is enabled, I need to define `self.model_config` as class `FuncConfig`, which is a new dataclass as the configuration including `functions`. As I need to add function objects into this `FuncConfig`, `FuncConfig` cannot be a frozen dataclass which prevents it from inheriting `ChatGPTConfig`. Thus `FuncConfig` will be a completely different class from `ChatGPTConfig`. 
Then the problem comes that `mypy` will report an error if I declare `self.model_config: FuncConfig = ...` and `self.model_config: ChatGPTConfig = ...` in the 2 cases of if-else statement (I remember it will be a duplicate. Thus, considering only the dictionary will essentially be made use by the backend model to call OpenAI API, I just kept this as a field of type `dict` and let it load values from the 2 different dataclasses.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -112,17 +145,65 @@ def __init__(
+                raise ValueError(
+                    "Function calling is enabled but no function is provided")
+
+            # Change the model to the corresponding function-enabled one
```
will do this.

MorphlingEd commented on camel/typing.py:
```diff
@@ -20,13 +20,16 @@ class RoleType(Enum):
 class ModelType(Enum):
     GPT_3_5_TURBO = "gpt-3.5-turbo"
     GPT_3_5_TURBO_16K = "gpt-3.5-turbo-16k"
+    GPT_3_5_TURBO_FUNC = "gpt-3.5-turbo-0613"
```
Cool! I will then remove the `FUNC` types later :)

MorphlingEd commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,105 @@
+    return ' '.join(sentences[:5])
+
+
+def search(entity: str) -> str:
```
That's true. Will do this.

MorphlingEd commented on camel/configs.py:
```diff
@@ -74,3 +74,71 @@ class ChatGPTConfig:
+
+
+@dataclass
+class FuncConfig:
```
🤔 My original idea is to inherit from `ChatGPTConfig` but an unfrozen dataclass cannot inherit a frozen one (`FuncConfig` should be unfrozen to make `functions` mutable). Could you further explain how an instance be put in `FuncConfig`? Is it keeping an instance of `ChatGPTConfig` and using an extra `to_dict`-like method for parsing it into a dictionary containing both fields from `ChatGPTConfig` and `functions` when needed (but I am not sure)?

Obs01ete replied:
You explained the right solution yourself very well!

dandansamax replied:
>  Is it keeping an instance of ChatGPTConfig and using an extra to_dict-like method for parsing it into a dictionary containing both fields from ChatGPTConfig and functions when needed (but I am not sure)?

Yes, I think this is better than the current solution. However, I think maybe we do not need to change `functions` during one chat? Is there any sceanario we have to do that?

MorphlingEd replied:
Right. Actually I did not mean to implement it for dynamically adding functions during the chat but it was just for parsing functions into message objects one-by-one and loading them to the list. Now I made another version to firstly parsing all functions and directly passing the whole list to the configuration class, which allows the class to be frozen and directly inherit from `ChatGPTConfig`.

Obs01ete commented on camel/messages/func_message.py:
```diff
@@ -66,12 +66,15 @@ def to_openai_assistant_message(self) -> OpenAIAssistantMessage:
-        msg_dict["function_call"] = {}
-        msg_dict["function_call"]["name"] = self.func_name
-        msg_dict["function_call"]["arguments"] = str(self.args)
+        msg_dict["function_call"] = {
```
This is better but it is stull mutation

Obs01ete commented on examples/test/test_ai_society_example.py:
```diff
@@ -14,9 +14,15 @@
     examples.ai_society.role_playing.main(ModelType.STUB)
+
+
+@pytest.mark.slow
```
Is it slow?

MorphlingEd replied:
emm this will cause a few rounds of chatting but I am not sure how to determine whether a test should be slow or not. Is there a threshold for this?

Obs01ete replied:
Let's have 3 seconds as a cutoff. How long does the test take?

MorphlingEd replied:
Then it should be marked as slow as this will last for more than 1 minute

MorphlingEd commented:
> @MorphlingEd Please address all comments and resolve them with "Resolve conversation" once done

will do this ASAP after carefully reading Tianqi's code and any further discussion with him. He has provided me with some further optimization on the code structure. Sorry for the late action as I am busy with packaging my stuffs in UK today but I will do this within today.

Obs01ete commented:
> > @MorphlingEd Please address all comments and resolve them with "Resolve conversation" once done
> 
> will do this ASAP after carefully reading Tianqi's code and any further discussion with him. He has provided me with some further optimization on the code structure. Sorry for the late action as I am busy with packaging my stuffs in UK today but I will do this within today.

No urge here, just a reminder

dandansamax commented on camel/agents/chat_agent.py:
```diff
@@ -233,45 +270,105 @@ def step(
+                        self.handle_stream_response(response, num_tokens)
+
+                # function calling disabled or stopped
+                if not isinstance(
```
> isinstance should not be there in the core code. If you need isinstance, you are doing something wrong.

dandansamax commented on camel/agents/chat_agent.py:
```diff
@@ -78,6 +79,31 @@ def to_openai_message(self):
+
 class ChatAgent(BaseAgent):
     r"""Class for managing conversations of CAMEL Chat Agents.
 
```
> In general, Any should not be there. Here we could make an empty ABC base class for `ChatGPTConfig` and other model configs. But a better solution would be injecting a pre constructed model backend. This will replace `model` and `model_config` and improve code cohesion. @lightaime

MorphlingEd replied:
@Obs01ete I have added an empty ABC class called `BaseConfig` in `configs.py` and modified the argument type to be `Optional[BaseConfig]`. I am wondering whether this is OK.

dandansamax reviewed (changes requested):
Comments from #219 by @Obs01ete.

Obs01ete commented on camel/configs.py:
```diff
@@ -11,12 +11,18 @@
 
 @dataclass(frozen=True)
-class ChatGPTConfig:
+class BaseConfig(ABC):
```
Good! See, you kept the type check on place and made dependency inversion to loosen the coupling of agents and model configs. The final step would be to move BaseConfig to a file of its own, but we can skip it for now.

Obs01ete commented on camel/societies/role_playing.py:
```diff
@@ -262,13 +278,22 @@ def init_agents(self, init_assistant_sys_msg: BaseMessage,
+            function_list=self.assist_funcs,
             **(assistant_agent_kwargs or {}),
         )
         self.assistant_sys_msg = self.assistant_agent.system_message
```
This is accessing the private field of a class. Please rework it into a property system_message or method get_system_message()

Obs01ete replied:
Okay, it is not in the diff.

Obs01ete commented on camel/societies/role_playing.py:
```diff
@@ -338,6 +363,9 @@ def init_chat(self) -> Tuple[BaseMessage, List[BaseMessage]]:
                      "Now start to give me instructions one by one. "
                      "Only reply with Instruction and Input."))
 
+        # if self.assist_func_enable:
```
Remove commented out code

Obs01ete commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,97 @@
+        ]
+
+        if any("may refer to:" in p for p in page):
+            search_wiki("[" + entity + "]")
```
Obs not initialized

Obs01ete commented on camel/functions/openai_function.py:
```diff
@@ -0,0 +1,75 @@
+        self.parameters = parameters or {"type": "object", "properties": {}}
+
+    @property
+    def parameters(self):
```
Normally, if you use properties, all private fields you name with underscore.

MorphlingEd replied:
I am not sure what to do with this piece of code as the private field `_parameters` has already been with underscore and `self.parameters` will call the getter method. I also applied this pattern to a new commit which adds setter and getter for `system_message` of `ChatAgent`. Is there anything wrong with this pattern?🤔

MorphlingEd replied:
@Obs01ete Hello, I just mentioned this conversation in the meeting and hope you could help with this question? :)

Obs01ete replied:
Okay, this implementation is good. self._parameters gets initialized in the setter.

Obs01ete commented on camel/agents/chat_agent.py:
```diff
@@ -113,9 +144,17 @@ def __init__(
+        # This is a very dirty implementation. We should consider to
+        # move the function execution part to another class.
+        self.func_enable = (function_list is not None)
+        if function_list is not None:
```
Else branch does not initialize self.func_dict

Obs01ete reviewed (changes requested):
All good, only need to refactor `step`

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -234,45 +280,104 @@ def step(
+                        self.handle_stream_response(response, num_tokens)
+
+                # function calling disabled or stopped
+                if (not self.func_enable) \
```
@Obs01ete I have modified the method `step` and now its main body has only 46 lines. The functionalities of message list truncation, directly returning when number of tokens exceeds, checking whether `step` should end and executing function are put into separate methods. 
But sorry they were not implemented as static as I found information and methods of `self` need to be frequently accessed.

MorphlingEd replied:
Sorry there occurs some problems with the Embodied Agent due to updates in APIs of `transformers`. Tianqi is working on it and creates another branch. I guess the tests can only be passed after successfully fixing the problem.

Obs01ete reviewed (changes requested):
Okay, this is better. Let's continue polishing the code.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -281,97 +281,73 @@ def step(
         messages = self.update_messages('user', input_message)
 
-        output_messages: Optional[List[BaseMessage]]
+        loop_msg: bool = True
```
@Obs01ete I think this is necessary due to the fact that function calling response and normal text response are different. Given our input, if there is a function needed to be call, the model will return a response with function arguments, function name, and **a `finish_reason` being "function_call" instead of the normal "stop"**. We then execute the function and pack the information of function name, arguments and execution results into messages. These messages are sent back to the model again, and model will repeat the process again but taking the last function call into consideration. The `step` should only be terminated when the model thinks there is no need to call any more function and return a response with the `finish_reason` being "stop". The last response will be a normal text message summarizing the previous function calls.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -281,97 +281,73 @@ def step(
         """
         messages = self.update_messages('user', input_message)
```
I can only find the `self.system_message: BaseMessage = system_message` in the initialization part but not in `step`? I am wondering why we need to make this change 🤔.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -281,97 +281,73 @@ def step(
-            openai_messages = [
-                record.to_openai_message() for record in messages
-            ]
+        while loop_msg:
```
done in the new commit

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -281,97 +281,73 @@ def step(
-            ]
+        while loop_msg:
+            # Format messages and get the token number
+            messages = self.truc_messages(messages)
```
done in the new commit

Obs01ete commented on camel/societies/role_playing.py:
```diff
@@ -70,6 +72,9 @@ class RolePlaying:
             task specify meta dict with. (default: :obj:`None`)
         output_language (str, optional): The language to be output by the
             agents. (default: :obj:`None`)
+        assist_funcs (Optional[List[OpenAIFunction]]): the function collection
```
Expand the name to full words. This shortening make the purpose unclear.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -234,45 +278,88 @@ def step(
             self.validate_model_response(response)
+
             if not self.model_backend.stream:
+                # flake8: noqa :E501
```
emm this one is to avoid PEP8 reporting error due to that this line is too long. I think the first line of this statement includes 4 returned values and it is not good to cut it off in the middle.

Obs01ete replied:
Haha. Please follow the rule of 80 symbols. Do not supress the warnings like this. If needed, split into several lines with brackets.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -213,7 +257,7 @@ def submit_message(self, message: BaseMessage) -> None:
         self.stored_messages.append(ChatRecord('assistant', message))
 
-    @retry(wait=wait_exponential(min=5, max=60), stop=stop_after_attempt(5))
+    # @retry(wait=wait_exponential(min=5, max=60), stop=stop_after_attempt(5))
```
Sorry I just forgot to remove the `#` notation when committing. This is just for easier debugging when developping because retrying takes time.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -113,9 +142,17 @@ def __init__(
-        self.model_config: ChatGPTConfig = model_config or ChatGPTConfig()
         self.message_window_size: Optional[int] = message_window_size
 
+        # This is a very dirty implementation. We should consider to
```
emm this was written by Tianqi previously as he thought it might be possible to move the function calling part separate. But after our discussion we realized it is quite hard to decouple this part from `ChatAgent` running logic (as mentioned in the reasons why we need to include a while-loop in method `step). The decoupling was later partly solved by packing the logic of function calling into a separate function `step_function_call` but the loop-based logic seems necessary to be kept here.

Obs01ete commented on camel/agents/chat_agent.py:
```diff
@@ -134,6 +168,37 @@ def reset(self):
+        self._system_message = value
+
+    @property
+    def func_enable(self) -> bool:
```
Please use full words for function names. Here I suggest `is_function_enabled`. In English it would better be `are_functions_enabled`. I am okay with either of these.

Obs01ete reviewed (changes requested):
Also please run code coverage and add tests to what you haven't covered.
https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md#coverage-

MorphlingEd commented:
@Obs01ete I added more tests for testing `ChatAgent`, `search_function` etc. Also modified the function names and removed `flake8` statement as you said. Now the coverage of `test_chat_agent` has achieved 99%. But I am not sure whether it is correct to test private method of a class explicitly (e.g. I directly tested `update_message` but not through `step`). I guess further I need to wait (and perhaps help) Tianqi for fixing the Python interpreter?

Obs01ete commented:
> @Obs01ete I added more tests for testing `ChatAgent`, `search_function` etc. Also modified the function names and removed `flake8` statement as you said. Now the statement coverage rate of class `ChatAgent` has achieved 99%. But I am not sure whether it is correct to test private method of a class explicitly (e.g. I directly tested `update_message` but not through `step`). I guess further I need to wait (and perhaps help) Tianqi for fixing the Python interpreter?

Thanks! You've added an impressive amount of tests. Testing a private method is fine. But you should never access private methods and fields in the core code. Yes, we do not use `self._private_method_of_field` notation yet, but we may move to it in the future. I am good with the MR in its current state. Please solve the failed tests and we can merge.

What is that issue with the failing test btw?

MorphlingEd commented:
> > @Obs01ete I added more tests for testing `ChatAgent`, `search_function` etc. Also modified the function names and removed `flake8` statement as you said. Now the statement coverage rate of class `ChatAgent` has achieved 99%. But I am not sure whether it is correct to test private method of a class explicitly (e.g. I directly tested `update_message` but not through `step`). I guess further I need to wait (and perhaps help) Tianqi for fixing the Python interpreter?
> 
> Thanks! You've added an impressive amount of tests. Testing a private method is fine. But you should never access private methods and fields in the core code. Yes, we do not use `self._private_method_of_field` notation yet, but we may move to it in the future. I am good with the MR in its current state. Please solve the failed tests and we can merge.
> 
> What is that issue with the failing test btw?

I have just checked it by myself and found the problem may be in the server of HuggingFace. When calling the HuggingFace tool (`image_generator` used in the current test), the server's response is 'Bad Gateway' and no valid information can be returned, resulting in decoding error locally.

Obs01ete commented:
@MorphlingEd I suggest you to fix the last issue so we are ready to merge

Obs01ete commented:
@lightaime please review this PR. Even though I am good with it, it is huge so please give it a look

lightaime commented:
Thanks for the great work @MorphlingEd @dandansamax @Obs01ete! I will look into it soon!

MorphlingEd commented:
> @MorphlingEd I suggest you to fix the last issue so we are ready to merge

Have done it and modified the example code to fit the new changes in `role_play`.

Obs01ete reviewed (approved):
Finally, it is in the mergable state. I approve but we will wait for the comments by @lightaime . Please do not merge before his comments.

lightaime reviewed (changes requested):
Thanks. Please do not merge it now. I will review it as soon as possible.

MorphlingEd commented:
> Thanks. Please do not merge it now. I will review it as soon as possible.

Sure.

Obs01ete commented on camel/agents/chat_agent.py:
```diff
@@ -91,20 +117,23 @@ class ChatAgent(BaseAgent):
 
         self.orig_sys_message: BaseMessage = system_message
-        self.system_message: BaseMessage = system_message
+        self.system_message = system_message
```
Actually the way it is implemented is nice. instead of the direct assignment to the frivate field, it goes through the property setter. I suggest not to make the change that you @lightaime propose. It works fine, I had checked it.

lightaime replied:
Oh, yeah, you are right. Please ignore this comment.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -234,45 +296,86 @@ def step(
+                of tokens exceeds the limit, None will replace the message
+                list
+        """
+        if (self.message_window_size is not None) and \
+                (len(messages) > self.message_window_size):
+            messages = \
+                [ChatRecord('system', self.system_message)] + \
+                messages[-self.message_window_size:]
+
+        openai_messages = \
+            [record.to_openai_message() for record in messages]
```
But I don't see why `/` should not be used...? I just found the modified code does not even look as well as that using `/`

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -350,6 +453,101 @@ def handle_stream_response(
+            tuple: a tuple containing the updated message history and the
+                struct containig information about this function call
+        """
+        choice = response["choices"][0]
```
Emm I just followed the typical way to do function calling and specified `n=1` in `FuncConfig`. Even when `n > 1`, I do not think selecting the first one with reason being `function_call` makes sense as sometimes there is indeed no need to call any function and the normal textual message with the highest probability is the best choice.

MorphlingEd commented on camel/configs.py:
```diff
@@ -74,3 +80,24 @@ class ChatGPTConfig:
+@dataclass(frozen=True)
+class FuncConfig(ChatGPTConfig):
+    r"""Defines the parameters for generating chat completions using the
+    OpenAI API with functions included. Note this is an unfrozen dataclass
+    allowing for dynamically adding functions into the functions field
```
Sorry this should be frozen and I forgot to remove this part of documentation when developping it.

MorphlingEd commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,104 @@
+        sentences += p.split('. ')
+    sentences = [s.strip() + '.' for s in sentences if s.strip()]
+
+    return ' '.join(sentences[:5])
```
Sorry this is not the number of available pages but the first 5 sentences in this page. I think it would be better to add another parameter to control the number of sentences to be returned.

MorphlingEd replied:
Well I thought hardcoding 5 sentences may be good... If I add another parameter to control the number of sentences, the model will try to adjust this and may not have a good control? Even if there are less than 5 pages (though such a Wiki page seems not existing), the range indexing can still work to return all sentences.

lightaime replied:
Sure, no prob

MorphlingEd commented on camel/societies/role_playing.py:
```diff
@@ -256,12 +272,22 @@ def init_agents(self, init_assistant_sys_msg: BaseMessage,
             output_language (str, optional): The language to be output by the
                 agents.
         """
+        if self.assistant_functions is not None:
+            assistant_config = FuncConfig(
+                functions=[
+                    func.as_dict() for func in self.assistant_functions
+                ], function_call="auto")
+        else:
+            assistant_config = None
```
I have discussed with Tianqi. His idea is that by principle the configuration should not be modified by `ChatAgent` and it should be prepared by `role_playing` before being passed to the agent class. I think it makes sense and wonder how you think of this.

lightaime replied:
`function_list` can be defined as a property method in `FuncConfig`.

MorphlingEd commented:
@lightaime Hi, I have just modified the code and resolved many of the conversations above. But there is still one thing not that clear: about the typing for the functions to be considered by the model's function calling. @dandansamax Tianqi searched for relevant parameter settings and found the argument specification should follow the JSON schema and the type settings are different with Python. For example, I must use `string` and `integer` instead of `str` and `int` otherwise errors will raise during the schema validation. So if I need to do float addition, I guess I need to define another function called `add_float` with parameter type being `number`, which is the general number type in json schema (https://json-schema.org/understanding-json-schema/reference/numeric.html#number). But I think we can add a more diverse set of functions later as this PR is mainly for implementing this function calling mechanism.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -234,45 +296,86 @@ def step(
-                num_tokens,
-            )
+            # Postprocess the response or execute function call
+            if self.step_end(finish_reasons[0]):
```
Then should I modify this if-statement into `if (not self.is_function_calling_enabled) or (finish_reasons[0] == 'stop'):`? (Though I previously used this one)

lightaime commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,107 @@
+                             "escape").encode("latin1").decode("utf-8")
+
+
+def get_page_obs(page: str) -> str:
```
The function name can be improved.

MorphlingEd replied:
I changed it to `get_page_abstract`. Is this better?

lightaime commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def search_wiki(entity: str) -> str:
+    r"""Search the entity in WikiPedia and return (the first sentences of)
```
```suggestion
    r"""Search the entity in WikiPedia and return (the first :obj:`5` sentences of)
```

lightaime commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,107 @@
+
+    Returns:
+        string: search result. If the page corresponding to the entity
+        exists, return the first 5 sentences in a string.
```
```suggestion
        exists, return the first :obj:`5` sentences in a string.
```

lightaime commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,107 @@
+        page (str): The fetched page.
+
+    Returns:
+        str: The concatenation of the first 5 sentences in the give page.
```
```suggestion
        str: The concatenation of the first :obj:`5` sentences in the given page.
```

lightaime commented on camel/functions/search_functions.py:
```diff
@@ -0,0 +1,107 @@
+            clean_str(div.get_text().strip()) for div in result_divs
+        ]
+        obs = (f"Could not find {entity}. "
+               f"Similar: {result_titles[:5]}.")
```
```suggestion
               f"Similar: {result_titles[:5] if len(result_titles)>5 else result_titles}.")
```

lightaime commented on camel/messages/func_message.py:
```diff
@@ -0,0 +1,100 @@
+    function-related messages.
+
+    Args:
+        func_name (Optional[str]): The name of the function used
+        args (Optional[Dict]): dictionary of arguments passed to the function
+        result (Optional[Any]): result of function execution
```
```suggestion
        func_name (Optional[str]): The name of the function used. (default: :obj:`None`)
        args (Optional[Dict]): The dictionary of arguments passed to the function. (default: :obj:`None`)
        result (Optional[Any]): The result of function execution. (default: :obj:`None`)
```

lightaime commented on camel/messages/func_message.py:
```diff
@@ -0,0 +1,100 @@
+
+    def to_openai_function_message(self) -> OpenAIMessage:
+        r"""Converts the message to an :obj:`OpenAIMessage` object
+        with the role being "function
```
```suggestion
        with the role being "function.
```

lightaime commented on camel/messages/func_message.py:
```diff
@@ -0,0 +1,100 @@
+
+        Returns:
+            OpenAIMessage: The converted :obj:`OpenAIMessage` object
+                with its role being "function"
```
```suggestion
                with its role being "function".
```

lightaime commented on camel/societies/role_playing.py:
```diff
@@ -256,12 +272,22 @@ def init_agents(self, init_assistant_sys_msg: BaseMessage,
                 agents.
         """
+        if self.assistant_functions is not None:
+            assistant_config = FuncConfig(
+                functions=[
+                    func.as_dict() for func in self.assistant_functions
+                ], function_call="auto")
```
Should we create a class method called `from_openai_function_list ` in `FuncConfig` to make it easier?

Like:
```
@classmethod
def from_openai_function_list(cls, function_list: List[OpenAIFunction], function_call="auto"):
    return cls(functions=[
                    func.as_dict() for func in function_list
                ], function_call=function_call)
```
So we can do:
```
assistant_config = FuncConfig.from_openai_function_list(self.from_openai_function_list)
```

MorphlingEd replied:
I think this implementation is the best as it does not break the current settings

lightaime commented on camel/utils/functions.py:
```diff
@@ -239,3 +240,67 @@ def download_tasks(task: TaskType, folder_path: str) -> None:
+    description and parameters.
+
+    Args:
+        func (Callable): the function to be parsed
```
```suggestion
        func (Callable): the Tunction to be parsed.
```

lightaime commented on camel/utils/functions.py:
```diff
@@ -239,3 +240,67 @@ def download_tasks(task: TaskType, folder_path: str) -> None:
+
+
+def parse_doc(func: Callable) -> Dict[str, Any]:
+    """
+    Parse the docstrings of a function to extract function name,
```
```suggestion
    r"""Parse the docstrings of a function to extract the function name,
```

lightaime commented on camel/utils/functions.py:
```diff
@@ -239,3 +240,67 @@ def download_tasks(task: TaskType, folder_path: str) -> None:
+    if len(sign_params) != len(required):
+        raise ValueError(f"Number of parameters in function signature "
+                         f"({len(sign_params)}) "
+                         f"does not match that in docstring ({len(required)})")
```
```suggestion
                         f"does not match that in docstring ({len(required)})".)
```

lightaime commented on camel/utils/functions.py:
```diff
@@ -239,3 +240,67 @@ def download_tasks(task: TaskType, folder_path: str) -> None:
+    for param in sign_params:
+        if param not in required:
+            raise ValueError(f"Parameter '{param}' in function signature"
+                             " is missing in the docstring")
```
```suggestion
                             " is missing in the docstring".)
```

lightaime commented on examples/function_call/role_playing_with_function.py:
```diff
@@ -0,0 +1,89 @@
+def main(model_type=None) -> None:
+    task_prompt = ("Estimate KAUST's current age by searching the year when "
+                   "it was founded, then add 10 more years to it. Assume the "
+                   "current year is 2023")
```
```suggestion
                   "current year is 2023.")
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def add_with_doc(a: int, b: int) -> int:
+    r"""Adds two numbers
```
```suggestion
    r"""Adds two numbers.
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+        b (integer): number to be added
+
+    Returns:
+        integer: the sum of the two numbers
```
```suggestion
        integer: The sum of the two numbers.
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def add_with_wrong_doc(a: int, b: int) -> int:
+    r"""Adds two numbers
```
```suggestion
    r"""Adds two numbers.
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+    r"""Adds two numbers
+
+    Args:
+        a (integer): number to be added
```
```suggestion
        a (integer): The first number to be added.
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+        a (integer): number to be added
+
+    Returns:
+        integer: the sum of the two numbers
```
```suggestion
        integer: The sum of the two numbers.
```

lightaime commented on test/functions/test_search_functions.py:
```diff
@@ -0,0 +1,61 @@
+    expected_output = (
+        "Erygia sigillata is a species of moth in the family Erebidae found "
+        "in Himachal Pradesh, Northern India.[2] The moth was officially "
+        "recognized and classified in 1892.. This Erebinae-related article "
```
Why this is two `.` after 1892?

MorphlingEd replied:
A good point! That is because the implementation will split a paragraph by `. ` and then concatenates them by `'.'.join(...)`. But the problem is the last sentence in a paragraph will still have a trailing '.'. I will remove this in the next commit

lightaime commented on test/agents/test_chat_agent.py:
```diff
@@ -82,6 +156,27 @@ def test_chat_agent_multiple_return_messages(n):
+
+    with pytest.raises(
+            RuntimeError,
+            match="Property msg is only available for a single message in msgs"
```
```suggestion
            match="Property msg is only available for a single message in msgs."
```

lightaime commented on test/agents/test_chat_agent.py:
```diff
@@ -137,3 +241,74 @@ def test_set_output_language():
+
+    user_msg = BaseMessage(role_name="User", role_type=RoleType.USER,
+                           meta_dict=dict(),
+                           content="Calculate the result of: 2*8-10")
```
```suggestion
                           content="Calculate the result of: 2*8-10.")
```

lightaime commented on test/functions/test_openai_function.py:
```diff
@@ -0,0 +1,82 @@
+def test_function_without_doc():
+    with pytest.raises(
+            ValueError,
+            match="Invalid function add_without_doc: no docstring provided"):
```
```suggestion
            match="Invalid function `add_without_doc`: no docstring provided."):
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+
+    with pytest.raises(
+            ValueError,
+            match=f"Invalid function {sub.__name__}: no docstring provided"):
```
The original error message could be improved.
```suggestion
            match=f"Invalid function `{sub.__name__}`: no docstring provided."):
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+def test_parse_doc_mismatch():
+
+    def mul(a, b, c):
+        """Multiplies three numbers
```
```suggestion
        r"""Multiplies three numbers.
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+        """Multiplies three numbers
+
+        Args:
+            a (int): number to be multiplied
```
```suggestion
            a (int): The first number to be multiplied.
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+
+        Args:
+            a (int): number to be multiplied
+            b (int): another number to be multiplied
```
```suggestion
            b (int): The second number to be multiplied.
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+        """Return the parameter
+
+        Args:
+            b (int): parameter to be returned
```
```suggestion
            b (int): The parameter to be returned.
```

lightaime commented on test/utils/test_parse_doc.py:
```diff
@@ -0,0 +1,79 @@
+
+    with pytest.raises(
+            ValueError, match=("Parameter 'a' in function signature"
+                               " is missing in the docstring")):
```
```suggestion
                               " is missing in the docstring")):
```
```suggestion
                               " is missing in the docstring.")):
```

lightaime commented:
> @lightaime Hi, I have just modified the code and resolved many of the conversations above. But there is still one thing not that clear: about the typing for the functions to be considered by the model's function calling. @dandansamax Tianqi searched for relevant parameter settings and found the argument specification should follow the JSON schema and the type settings are different with Python. For example, I must use `string` and `integer` instead of `str` and `int` otherwise errors will raise during the schema validation. So if I need to do float addition, I guess I need to define another function called `add_float` with parameter type being `number`, which is the general number type in json schema (https://json-schema.org/understanding-json-schema/reference/numeric.html#number). But I think we can add a more diverse set of functions later as this PR is mainly for implementing this function calling mechanism.

Totally agree with this. Thanks!

Obs01ete commented:
Why are gradio app tests failing?
https://github.com/camel-ai/camel/actions/runs/5659852994/job/15334286171?pr=213

MorphlingEd commented:
> Why are gradio app tests failing? https://github.com/camel-ai/camel/actions/runs/5659852994/job/15334286171?pr=213

I am not sure.. The test fails just after Guohao merges master into this branch🤔but the error message seems to indicate problems not related to this branch.

lightaime commented on camel/societies/role_playing.py:
```diff
@@ -256,12 +272,22 @@ def init_agents(self, init_assistant_sys_msg: BaseMessage,
             init_assistant_sys_msg,
+            model_config=assistant_config,
             output_language=output_language,
+            function_list=self.assistant_functions,
```
One last concern is do we want to `function_list` as input args? Since it is always should be always the same as the `function_list` used to define `model_config`. A better way may be to get the `function_list` from `model_config` inside `ChatAgent`. @MorphlingEd @Obs01ete @dandansamax WDYT?

MorphlingEd replied:
I am not sure whether I correctly understand what you mean. But I remember we previously discussed whether we need to dynamically add new functions during the execution but it seems currently we do not need this (at least we do not come up with such an application scene?). So we designed like this to fix the function list in the `model_config` object. I am wondering whether you have any suggestion on this.

lightaime commented:
BTW for the Wikipedia search. This is a package for it: https://pypi.org/project/wikipedia. Maybe we can use it in the future.

lightaime reviewed (commented):
Left some comments on renaming.

Obs01ete reviewed (approved):
LGTM at this stage. I approve, but please wait for the approval by Guohao

lightaime commented on camel/agents/chat_agent.py:
```diff
@@ -350,6 +457,94 @@ def handle_stream_response(
+        func = self.func_dict[func_name]
+
+        args = choice["message"]["function_call"]["arguments"]
+        args = json.loads(args.replace("\'", "\""))
```
Just want to understand. Why this replace is needed?

MorphlingEd replied:
@lightaime That is because `json` only supports `"` but not `'` and that's why I firstly implemented using `eval`. But after replacing, `json.loads` can correctly be executed so I just used this.

lightaime reviewed (commented):
Thanks for the awesome PR @MorphlingEd @dandansamax @Obs01ete! I will go ahead and merge it.
